samples: subsys: Add MediaPipe camera to display sample

ngphibang opened the pull request:
This PR adds a video pipeline sample using **libMP** which can leverage all current video samples. Moreover, thanks to libMP, the pipeline could be easily extended beyond its current scope towards a very complex one without increasing (much) the code complexity.

This PR depends on:

- https://github.com/zephyrproject-rtos/zephyr/pull/98514
- https://github.com/zephyrproject-rtos/zephyr/pull/108344
- https://github.com/zephyrproject-rtos/zephyr/pull/108345

Only the last commit should be reviewed

Currently, the pipeline sample is equivalent and even more advanced than the capture sample (thanks to the video transform element). It can leverage all the 4 video samples (capture, capture-to-lvgl, tcpserversink, uvc) and even more in the future. Once it is done, we can deprecate current non-libMP Zephyr video samples except the very basic one (e.g. the capture sample to demonstrate the video / display API usage).

TODO:

- Validate on all supported HW. Currently, only validated on native sim and i.MX RT1170
- Add **libMP's tcpserversink element (and together with some minor changes in STMicroelectronic's venc and jpeg drivers)** so that we can leverage tcpserversink sample with support for jpeg and h264 encoders.
- Add libMP's LVGL element to do some extra overlay so that we can leverage capture-to-lvgl sample (and more than that).
- Add libMP's **uvcsink** element so that we can leverage uvc device sample

- **_Use devicetree to describe the libMP pipeline topology_** and have a unique common code to parse and set up the pipeline (using DT macros).
This way, no need to add boilerplate code for each application (maybe zero code if app does not need to do other things) and avoid tons of "if" when there are small modifs between boards on the pipeline (like different properties setting, add some more elements, etc.). I see devicetree fit very well on this purpose (each element then has a DT binding .yaml to describe all its properties, etc.), except **it violates the concept of "DT is to describe HW"**. But if we find a way to separate it from the real HW devicetree to avoid pollution but still be able to reuse DT parse / generate tools / script so it will be OK ?


```
/pipeline {
    compatible = "zmp,pipeline";
    id = 0;

    cam {
        compatible = "zmp,zvid_src";
        id = 1;
        num_bufs = 100;
        crop_width = 640;
        crop_height = 480;
        hflip = 1;
        port {
            cam_src: endpoint {
                remote-endpoint-label = "capsfilter_sink";
            };
        };
    };

    capsfilter {
        compatible = "zmp,capsfilter";
        id = 2;
        pixel_format = "RGBP"
        image_width = 320;
        image_height = 240;
        ports {
            #address-cells = <1>;
            #size-cells = <0>;
            port@0 {
                capsfilter_sink: endpoint {
                    remote-endpoint-label = "cam_src";
                };
            };
            port@1 {
                capsfilter_src: endpoint {
                    remote-endpoint-label = "vidtrans_sink";
                };
            };
        }
    };

    vidtrans {
        compatible = "zmp,zvid_transform";
        id = 3;
        video_cid_rotate = 270;
        ports {
            #address-cells = <1>;
            #size-cells = <0>;
            port@0 {
                vidtrans_sink: endpoint {
                    remote-endpoint-label = "capsfilter_src";
                };
            };
            port@1 {
                vidtrans_src: endpoint {
                    remote-endpoint-label = "disp_sink";
                };
            };
        }
    };

    disp {
        compatible = "zmp,zdisp_sink";
        id = 4;
        port {
            disp_sink: endpoint {
                remote-endpoint-label = "vidtrans_src";
            };
        };
    };
};
```

josuah commented on samples/drivers/video/pipeline/src/main.c:
```diff
@@ -0,0 +1,156 @@
+	ret = mp_object_set_properties(
+		MP_OBJECT(source),
+		COND_CODE_0(CONFIG_PROP_NUM_BUFS, (), (PROP_NUM_BUFS, CONFIG_PROP_NUM_BUFS,))
+				 COND_CODE_0(CONFIG_VIDEO_SOURCE_CROP_WIDTH, (), (PROP_ZVID_CROP, &crop,))
```
it's possible to wrap some code in `/* clang-format off */` and `/* clang-format on */` in case it gives something too odd, to keep some manual formatting.

josuah reviewed (commented):
A few low hanging fruits, in the meantime I can gather time for a more complete review.

ngphibang commented:
Rebased on latest main (which has some recent changes in video).

josuah commented on samples/drivers/video/pipeline/src/main.c:
If you preferred `#if`/`#endif`:
```c
	if (!mp_element_link(
		MP_ELEMENT(&vid_src),
#if (CONFIG_MP_CAPSFILTER)
		MP_ELEMENT(&caps_filter),
#endif
#if (DT_HAS_CHOSEN(zephyr_videotrans))
		MP_ELEMENT(&vid_trans),
#endif
		MP_ELEMENT(&disp_sink),
		NULL)) {
```

josuah replied:
Then maybe it plays well with clang-format

josuah commented on samples/drivers/video/pipeline/src/main.c:
It might be possible to use `#if`/`#endif` if preferred:

```c
	ret = mp_object_set_properties(MP_OBJECT(&vid_src),
#if (CONFIG_PROP_NUM_BUFS)
		PROP_NUM_BUFS, CONFIG_PROP_NUM_BUFS,
#endif
#if (CONFIG_VIDEO_SOURCE_CROP_WIDTH)
		PROP_ZVID_CROP, &crop,
#endif
#if (CONFIG_VIDEO_CTRL_HFLIP)
		VIDEO_CID_HFLIP, CONFIG_VIDEO_CTRL_HFLIP,
#endif
#if (CONFIG_VIDEO_CTRL_VFLIP)
		VIDEO_CID_VFLIP, CONFIG_VIDEO_CTRL_VFLIP,
#endif
		PROP_LIST_END);
```

josuah commented on samples/drivers/video/pipeline/src/main.c:
If wanting to rely on `#if`/`#endif`:
```c
	struct mp_caps *caps = mp_caps_new(MP_MEDIA_VIDEO,
#if (CONFIG_VIDEO_FRAME_WIDTH)
		MP_CAPS_IMAGE_WIDTH, MP_TYPE_UINT, CONFIG_VIDEO_FRAME_WIDTH,
#endif
#if (CONFIG_VIDEO_FRAME_HEIGHT)
		MP_CAPS_IMAGE_HEIGHT, MP_TYPE_UINT, CONFIG_VIDEO_FRAME_HEIGHT,
#endif
#if (CONFIG_VIDEO_FRAME_RATE)
		MP_CAPS_FRAME_RATE, MP_TYPE_UINT_FRACTION, CONFIG_VIDEO_FRAME_RATE, 1,
#endif
		MP_CAPS_END);
```

josuah commented on samples/drivers/video/pipeline/src/main.c:
If wanting to rely on `#if`/`#endif`:
```c
	if (!mp_bin_add(MP_BIN(&pipe),
		MP_ELEMENT(&vid_src),
#if (CONFIG_MP_CAPSFILTER)
		MP_ELEMENT(&caps_filter),
#endif
#if (DT_HAS_CHOSEN(zephyr_videotrans))
		MP_ELEMENT(&vid_trans),
#endif
	 	MP_ELEMENT(&disp_sink),
		NULL)) {
```

josuah commented on samples/drivers/video/pipeline/src/main.c:
If wanting to rely on `#if`/`#endif`:

```c
	ret = mp_object_set_properties(
		MP_OBJECT(&vid_trans),
#if (CONFIG_VIDEO_ROTATION_ANGLE)
		VIDEO_CID_ROTATE, CONFIG_VIDEO_ROTATION_ANGLE,
#endif
		PROP_LIST_END);
```

josuah reviewed (commented):
Just a small pass about use of conditional macros.

libMP is having plenty of macros already, and combining this with `COND_CODE` and `IF_ENABLED()` requires very good knowledge of both libMP and Zephyr.

To make it more familiar to non-Zephyr-macro developers (i.e. v4l2 devs, devs using Zephyrs extensively but not macros all the time) and non-Gstreamer developers, how about using `#if`/`#endif` which everybody will know?

ngphibang committed:
mp: Add core tests

Add build-only, unit and mock pipeline tests for the MP core.

The mock pipeline is composed of a fake source, a transform and
a sink to verify the whole core framework behavior such as
pipeline creation, caps negotiation and data flow.

Assisted-by: Claude:claude-opus-4.6
Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>
Signed-off-by: Trung Hieu Le <trunghieu.le@nxp.com>

ngphibang committed:
mp: Introduce Media Pipe (MP) subsystem

MP (Media Pipe) is a lightweight GStreamer-like multimedia framework
for Zephyr. MP reuses many concepts from GStreamer, such as elements,
pads, caps negotiation, and buffer negotiation and adopts a pipeline-
based architecture that decomposes multimedia processing into discrete,
interconnected elements.

It aims to simplify the development of multimedia applications by
providing simple and stable APIs for users to rapidly create their
specific applications, i.e., users simply select the built-in elements
and plugins suited to their purpose to construct a pipeline, and it
just works. This design promotes modularity, reusability, and efficient
resource management (e.g., zero-copy data flow). Moreover, the APIs
are generic enough so that application code can remain unchanged even
as MP evolves.

MP also features a highly modular, inheritance-based architecture
inspired by GStreamer, ensuring modularity, scalability, and
maintainability. For example, new custom elements can be easily added
by extending existing elements, without requiring modifications to the
core components. Plugins are decentralized from the core structures,
allowing seamless extension without altering the core framework.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>
Signed-off-by: Trung Hieu Le <trunghieu.le@nxp.com>

ngphibang committed:
mp: Add utils

Add the utils helper. These are optional, reusable utilities built on
top of the subsys to simplify application development.

The utils currently includes mp_player, a small pipeline controller
that drives a pipeline through its states and exposes simple
play/pause/stop/replay/quit controls (usable from a shell). More
utilities may be added in the future.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
mp: Add base plugin

Add the base plugin for the MP subsystem. This plugin provides
generic, reusable elements like:
- queue: pipeline-level threading element
- tee: pipeline branching element
- capsfilter: caps enforcement element

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
mp: Add video plugin

Add the vid (Zephyr Video) plugin for the MP subsystem. This plugin
provides video-specific elements that interface with Zephyr's video
subsystem, enabling building video capture and processing pipelines
using Zephyr video devices, e.g. camera, m2m devices

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
mp: Add display plugin

Add the disp (Zephyr Display) plugin for the MP subsystem. This
plugin provides display output elements that interface with Zephyr's
display subsystem, enabling building video display pipelines that
output processed frames to physical displays.

The plugin includes a display sink element that renders video frames
to a Zephyr display device, supporting partial frame updates and
configurable display regions.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
mp: samples: Add camera to display sample

Add the cam_disp sample application demonstrating how to build a
camera-to-display pipeline using the MP subsystem. This sample
captures video frames from a camera device using the vid plugin and
renders them on a display using the disp plugin, showcasing
real-time video preview functionality.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>
